### PR TITLE
Fix buffer overrun while parsing malformed JSON hex escape sequence

### PR DESCRIPTION
**Pull request by martijnthe**
Fixes https://github.com/jerryscript-project/jerryscript/issues/2200

JerryScript-DCO-1.0-Signed-off-by: Martijn The martijn.the@intel.com

### CODE REVIEW

**Comment by martijnthe** [comment]
cc @DanielBallaSZTE @zherczeg @renatahodovan

**Comment by martijnthe** [comment]
@akosthekiss Any idea what's wrong with the sign-off signature?
What CI reports is exactly what's at the end of the commit message...

**Comment by zherczeg** [comment]
What do you think about #2194 ?

**Comment by martijnthe** [comment]
@zherczeg ah, I had not seen there was already a PR for it.

I think #2194 is still not quite correct. I think the comparison should include the count for 1 char for the `u` and 4 chars for the hex value. It's also missing a test case as you pointed out.

In my PR I also replaced the various uses of magic number `4` with a `#define`, which is a bit better IMO.

**Comment by martijnthe** [comment]
> @akosthekiss Any idea what's wrong with the sign-off signature?
What CI reports is exactly what's at the end of the commit message...
Should there be a newline at the end perhaps?
Or is it choking on the é?

OK, the problem is the `é` was represented differently in the commit message text vs author field...

**Comment by robertsipka** [comment]
@martijnthe : Yes you are right, I've missed to increase with 1. I also strongly support the removal of magic constants. But the mistake that was made earlier to fix this kind of issue is still there. I've suggest you to modify the condition as in #2194 instead of adding a new one before.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c [hunk 1]
```diff
@@ -182,13 +187,17 @@ ecma_builtin_json_parse_string (ecma_json_token_t *token_p) /**< token argument
         }
         case LIT_CHAR_LOWERCASE_U:
         {
+          if ((end_p - current_p < ECMA_JSON_HEX_ESCAPE_SEQUENCE_LENGTH + 1)) {
+            return;
+          }
+
           ecma_char_t code_unit;
           if ((end_p - current_p >= 2) && !(lit_read_code_unit_from_hex (current_p + 1, 4, &code_unit)))
```
Comment (robertsipka) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: You should use ECMA_JSON_HEX_ESCAPE_SEQUENCE_LENGTH here too.

Comment (martijnthe) [reply jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: Oh whoops, I intended to remove that condition there. It would be redundant otherwise.

**Comment by martijnthe** [comment]
@robertsipka OK, fixed it.

Should I also create a `#define` to replace the magic `1` (the length of the `u`)?
What do you think?

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c [hunk 2]
```diff
@@ -182,13 +187,17 @@ ecma_builtin_json_parse_string (ecma_json_token_t *token_p) /**< token argument
         }
         case LIT_CHAR_LOWERCASE_U:
         {
+          if ((end_p - current_p < ECMA_JSON_HEX_ESCAPE_SEQUENCE_LENGTH + 1)) {
+            return;
+          }
+
           ecma_char_t code_unit;
-          if ((end_p - current_p >= 2) && !(lit_read_code_unit_from_hex (current_p + 1, 4, &code_unit)))
+          if (!(lit_read_code_unit_from_hex (current_p + 1, 4, &code_unit)))
```
Comment (robertsipka) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: You've forgot to replace 4 with ECMA_JSON_HEX_ESCAPE_SEQUENCE_LENGTH here.

Comment (robertsipka) [reply jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: I do not think it is necessary. LGTM (informally) after this minor fix.

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c [hunk 3]
```diff
@@ -182,13 +187,17 @@ ecma_builtin_json_parse_string (ecma_json_token_t *token_p) /**< token argument
         }
         case LIT_CHAR_LOWERCASE_U:
         {
+          if ((end_p - current_p < ECMA_JSON_HEX_ESCAPE_SEQUENCE_LENGTH + 1)) {
+            return;
```
Comment (zherczeg) [inline jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: what about <= ? Then the +1 is unnecessary.

Comment (martijnthe) [reply jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: Done

File: jerry-core/ecma/builtin-objects/ecma-builtin-json.c (hunk 2 above)
Comment (martijnthe) [reply jerry-core/ecma/builtin-objects/ecma-builtin-json.c]: Done

**Comment by martijnthe** [comment]
OK, I also satisfied the linter now and the CI build is ✅

**Pull request merged**